### ISSUE TITLE
Errors should be instance of Error, not string

### ISSUE THREAD

**Issue by WoLfulus** [nats-io/node-nats#104]
This gets really hard to handle exceptions without proper error codes, for example, I'd like to shutdown the application if there's a ETIMEDOUT error, and keep running on other codes. 

Check if the error string contains "ETIMEDOUT" isn't an option.

**Comment by aricart** [comment]
What error are you getting exactly? This is likely an error coming from the socket. Where are you getting it. Can you share the stack trace?

**Comment by WoLfulus** [comment]
I mean, the lib itself emits errors as strings, not Error instances.

https://github.com/nats-io/node-nats/blob/master/lib/nats.js#L310

I can't rely on analysing string contents to check what errors are happening.

Example: 

```
this.emit('error', new SecureConnectionRequiredError());
```

**Comment by aricart** [comment]
Let me see what  I can do. This change will create a backwards compatible issues for everyone else if they are relying on the string.

**Comment by aricart** [comment]
@WoLfulus go ahead a see if the PR fits for you. In the case of the connection issue, instead of wrapping the error or adding additional properties, I am just bubbling the error up without any kind of augmentation. All errors generated by the library now have constants as well to aid in testing the errors. If there are any other issues, please feel free to reopen or create another issue.  PR will be merged once additional eyeballs land on it.

**Comment by WoLfulus** [comment]
Where does the `code` of the error goes? Error class receives only one parameter on the constructor. The second one is lost.

**Comment by aricart** [comment]
This is what I just did the source for Error gave me the impression that the error id was exposed. I implemented exactly what you propose. The NatsError also provides a `code`, `chainedError` so you can compare codes. If there's a connection error you can further inspect the `chainedError` for the reason. Thanks for the catch.

**Comment by WoLfulus** [comment]
Perfect! Thanks :)
